// @aws-amplify/ui-components/dist/esm-es5/amplify-greetings.entry.js
import { r as registerInstance, h } from './index-a93ff41e.js';
import '@aws-amplify/core';
import './auth-types-78df304e.js';
import './Translations-06ef20e7.js';
import './constants-6835ae6a.js';
import '@aws-amplify/auth';
import { d as dispatchAuthStateChangeEvent } from './helpers-910cfbb7.js';
var amplifyGreetingsCss = ":host{--background-color:var(--amplify-white);--border-color:var(--amplify-light-grey);--font-family:var(--amplify-font-family)}.greetings{position:relative;border:1px solid transparent;background-color:var(--background-color);border-color:var(--border-color);padding:10px;font-family:var(--font-family)}amplify-sign-out{margin-left:1rem}";
var AmplifyGreetings = /** @class */ (function () {
    function AmplifyGreetings(hostRef) {
        registerInstance(this, hostRef);
        /** Username displayed in the greetings */
        this.username = null;
        /** Logo displayed inside of the header */
        this.logo = null;
        /** Auth state change handler for this component */
        this.handleAuthStateChange = dispatchAuthStateChangeEvent;
    }
    AmplifyGreetings.prototype.render = function () {
        return (h("header", { class: "greetings" }, this.logo && h("span", null, this.logo), h("amplify-nav", null, this.username && (h("slot", { name: "greetings-message" }, h("span", null, "Hello, ", this.username))), h("amplify-sign-out", { handleAuthStateChange: this.handleAuthStateChange }))));
    };
    return AmplifyGreetings;
}());
AmplifyGreetings.style = amplifyGreetingsCss;
export { AmplifyGreetings as amplify_greetings };
